File: Kordia_backend/app/api/routes/stream.py
"""
Rutas de streaming
"""
from fastapi import APIRouter, HTTPException, Depends
from fastapi.responses import StreamingResponse, Response
import aiosqlite
import aiohttp
from app.schemas.responses import StreamResponse
from app.services.youtube_service import YouTubeService
from app.services.cache_service import CacheService
from app.database.repositories.stream_cache import StreamCacheRepository
from app.dependencies import get_youtube_service, get_cache_service, get_db

router = APIRouter(prefix="/stream", tags=["stream"])


@router.get("/{ytid}", response_model=StreamResponse)
async def get_stream_url(
    ytid: str,
    youtube_service: YouTubeService = Depends(get_youtube_service),
    cache_service: CacheService = Depends(get_cache_service),
    db: aiosqlite.Connection = Depends(get_db)
):
    """
    Obtener URL de stream de audio para un video de YouTube
    
    - **ytid**: ID del video de YouTube
    
    Returns URL del stream con indicador de si proviene del caché
    """
    try:
        # Crear repositorio
        cache_repo = StreamCacheRepository(db)
        
        # Verificar caché
        cached_url = await cache_service.get_stream_url(ytid, cache_repo)
        if cached_url:
            return StreamResponse(ytid=ytid, url=cached_url, cached=True)
        
        # Obtener nueva URL
        url = await youtube_service.get_stream_url(ytid)
        
        if not url:
            raise HTTPException(
                status_code=404,
                detail="No se pudo obtener URL de stream"
            )
        
        # Guardar en caché
        await cache_service.save_stream_url(ytid, url, cache_repo)
        
        return StreamResponse(ytid=ytid, url=url, cached=False)
    
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error obteniendo stream: {str(e)}"
        )

from fastapi import BackgroundTasks
import shutil

@router.get("/proxy/{ytid}")
async def proxy_stream(
    ytid: str,
    background_tasks: BackgroundTasks,
    youtube_service: YouTubeService = Depends(get_youtube_service),
    cache_service: CacheService = Depends(get_cache_service),
    db: aiosqlite.Connection = Depends(get_db)
):
    """
    Proxy directo del stream de audio para guardado local en clientes móviles.
    Usa yt-dlp para descargar rápido a disco y FileResponse para transferir C++-speed bytes,
    limpiando usando background tasks.
    """
    try:
        import tempfile
        import os
        from fastapi.responses import FileResponse
        
        # Crear un directorio que no se auto-borra inmediatamente
        tmpdir = tempfile.mkdtemp()
        temp_path_base = os.path.join(tmpdir, ytid)
        
        # fast_download_audio descarga el raw a disco SIN pasar por FFmpeg (mucho más rápido)
        ext = await youtube_service.fast_download_audio(ytid, temp_path_base)
        downloaded_file = f"{temp_path_base}.{ext}"
                
        if not os.path.exists(downloaded_file):
            shutil.rmtree(tmpdir, ignore_errors=True)
            raise HTTPException(status_code=500, detail="Error de descarga temporal rápida")
        
        # Aseguramos que se borre UNA VEZ TERMINADO DE SERVIR usando BackgroundTasks
        background_tasks.add_task(shutil.rmtree, tmpdir, ignore_errors=True)
        
        # FileResponse usa los binarios asincrónicos de C más rápidos (sendfile)
        return FileResponse(
            path=downloaded_file,
            media_type="audio/mp4",
            filename=f"{ytid}.{ext}"
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: Kordia_backend/src/routes/stream.ts
import type { FastifyInstance } from 'fastify';
import { mkdtempSync, existsSync } from 'fs';
import { rm } from 'fs/promises';
import { join } from 'path';
import { tmpdir } from 'os';
import { YouTubeService } from '../services/youtube.js';
import { CacheService } from '../services/cache.js';
import { getDb } from '../database/connection.js';
import { StreamCacheRepository } from '../database/stream-cache.js';
import { config } from '../config/env.js';

const youtube = new YouTubeService();
const cache = new CacheService();

export async function streamRoutes(app: FastifyInstance): Promise<void> {
  app.get('/stream/:ytid', async (req, reply) => {
    const { ytid } = req.params as { ytid: string };
    try {
      const db = getDb();
      const repo = new StreamCacheRepository(db);
      const cached = cache.getStreamUrl(ytid, repo);
      if (cached) return { ytid, url: cached, cached: true };

      const url = youtube.getStreamUrl(ytid);
      if (!url) return reply.code(404).send({ error: 'No se pudo obtener URL de stream' });

      cache.saveStreamUrl(ytid, url, repo);
      return { ytid, url, cached: false };
    } catch (e: any) {
      throw { statusCode: 500, message: `Error obteniendo stream: ${e.message}` };
    }
  });

  app.get('/stream/proxy/:ytid', async (req, reply) => {
    const { ytid } = req.params as { ytid: string };
    try {
      const tmpdirPath = mkdtempSync(join(tmpdir(), 'kordia-'));
      const basePath = join(tmpdirPath, ytid);
      const ext = youtube.fastDownloadAudio(ytid, basePath);
      const filePath = `${basePath}.${ext}`;

      if (!existsSync(filePath)) {
        await rm(tmpdirPath, { recursive: true, force: true });
        return reply.code(500).send({ error: 'Error de descarga temporal' });
      }

      reply.header('Content-Type', 'audio/mp4');
      reply.header('Content-Disposition', `attachment; filename="${ytid}.${ext}"`);
      const stream = require('fs').createReadStream(filePath);
      reply.send(stream);

      // Cleanup after response
      reply.then(() => rm(tmpdirPath, { recursive: true, force: true }).catch(() => {}), () => {});
    } catch (e: any) {
      throw { statusCode: 500, message: e.message };
    }
  });
}
